keps/0022-diagnostics-bundle.md

```diff
@@ -0,0 +1,423 @@
+  Instance
+- Current plan status, if one exists, or the KUDO Instance
+- Information about the KUDO instance's Operator and OperatorVersion
+- Logs for the KUDO controller manager
+- Describe for the KUDO controller manager resources
+- RBAC resources that are applicable to the KUDO controller manager
+- Current settings and version information for KUDO
+- Status of last preflight check run.
+
+Operator developer experience, then, focuses on customizing diagnostics
+information to gather information about the running application. The following
+forms are available, subject to change over time:
+
+- **Copy**: Copy a file out of a running pod. This is useful for non-stdout
+  logs, configuration files, and other artifacts generated by an application.
+  Higher level resources can also be used, which will copy the file on all pods
+  selected by that resource.
+- **Command**: Run a command on a running pod and copy the stdout. Higher level
+  resources can also be used, which will run the command on all pods selected by
+  that resource.
+- **Task**: Run a KUDO task and copy the stdout and other arbitrary files.
+- **HTTP**: Make an HTTP request from the KUDO controller manager to a named
+  service and port and copy the result of the request.
+
+While some of these are redundant (HTTP can be a command or job), the intent
+is to provide a high level experience where possible so that operator developers
+don't necessarily need to maintain a `curl` container as part of their
+application stack.
+
+Operator-defined diagnostics collection is defined in a new `diagnostics.bundle.resources`
+key in `operator.yaml`:
+
+```yaml
+diagnostics:
+  bundle:
+    resources:
+      - name: Zookeeper Configuration File
+        key: "zookeeper-configuration"
+        kind: Copy
+        spec:
+          path: /opt/zookeeper/server.properties
+          objectRef:
+            kind: StatefulSet # Runs on ALL pods in the statefulset
+            name: "{{ .InstanceName }}-zookeeper"
+      - name: DNS information for running pod
+        key: "dns-information"
+        kind: Command
+        spec:
+          command: # Can be string or array
+            - nslookup
+            - google.com
+          objectRef:
+            kind: Pod
+            name: "{{ .InstanceName }}-zookeeper-0"
+    filters:
+      - name: Authentication information
+        spec:
+          regex: "^host: %w+$"
+```
+
+This key is **OPTIONAL**. Default diagnostics collection will happen regardless
+of the `diagnostics.bundle` key's presence. Note, moving to a graph-based engine
+for KUDO will make selecting of resources much easier, rather than having to
+use magical strings with templates. Future iterations of this will reduce the
+complexity of selecting resources to run commands and files on.
+
+Steps in a bundle run serially. To prevent the KUDO controller manager from
+crashing, the collector process runs in another pod as a job. **TODO**:
```

Comment:
Ok, I've taken another close look at what data we want to collect and **all** of it is available from the CLI. We have a mix of `k get/describe/logs/events` and `k exec/cp` for executing commands and saving stdout/copy out files from the pods. Frankly, all of that could be a rather simple kudo-cli command.

If our past taught us anything, then we know that logs from the big clusters can easily exceed gigabytes (even gzipped). Handling them, either in the Manager or in a separate Pod, will require arbitrary size PVCs?

Reply:
A user of an operator collecting these diagnostics might not have RBAC to run all of these locally, ESPECIALLY exec/copy - which can present security issues.

Reply:
That said, I'm open to this approach too. It simplifies a lot of things. :)

Reply:
Could things like resumable/retried diagnostic-fetching (`wget -C`) be feasibly implemented as a set of operations under a single CLI command?

> Handling them, either in the Manager or in a separate Pod, will require arbitrary size PVCs?

That's a good point. It should at a minimum be called out in documentation if that's the path taken, imo.

Reply:
> A user of an operator collecting these diagnostics might not have RBAC to run all of these locally, ESPECIALLY exec/copy - which can present security issues.

Fair point. Though I expect most users have access to "their stuff" or having some kind of admin on the line with the right access. It seems like a small price to pay for a **massive** implementation  simplification 🤷‍♀

Reply:
> Could things like resumable/retried diagnostic-fetching (`wget -C`) be feasibly implemented as a set of operations under a single CLI command?

That's the easiest way to implement them IMHO. Everything else adds layers of complexity: extra image to maintain, extra Pod/Job to deploy, extra things that can failover to think about...

Reply:
> A user of an operator collecting these diagnostics might not have RBAC to run all of these locally, ESPECIALLY exec/copy - which can present security issues.

Oooh.. this really opens a can of worms.

This seems to imply that we treat the controller as a privileged entity, and therefore treat `OperationVersion`s as trusted resources (because their diagnostics spec can prescribe whatever), and `Instances` as untrusted ones. However I do not think we mention this anywhere in the operator developer docs - for example that the developer should treat parameter values as untrusted input.

I think for now we should do something like:
1. say in the KEP that it's unspecified where the diagnostics collection runs,
1. actually run it from the CLI for now, and require RBAC to be opened up as needed

And in parallel we can start thinking about KUDO security model more broadly.